cmd/tink-worker/main.go

```diff
@@ -2,18 +2,56 @@ package main
 
 import (
 	"fmt"
+	"math/rand"
 	"os"
+	"time"
 
+	"github.com/tinkerbell/tink/cmd/tink-worker/client"
 	"github.com/tinkerbell/tink/cmd/tink-worker/cmd"
+	"github.com/tinkerbell/tink/protos/workflow"
+	"google.golang.org/grpc"
 )
 
 func main() {
-	cmdlineFlags, err := cmd.CollectCmdlineFlags(os.Args[1:])
+	// parse and validate command-line flags and required env vars
+	flagEnvSettings, err := cmd.CollectFlagEnvSettings(os.Args[1:])
 	if err != nil {
 		fmt.Println("Error:", err)
 		os.Exit(1)
 	}
 
-	// Remove me: this is just here to prevent linter issues
-	fmt.Printf("Version flag is %v\n", cmdlineFlags.Version)
+	// establish connection to tink server
+	creds, err := client.ObtainTinkServerCreds(flagEnvSettings.TinkServerURL)
+	if err != nil {
+		fmt.Println("Error obtaining tink server TLS credentials:", err)
+		os.Exit(1)
+	}
+	// Here we retry failed connections to tink-server using a randomized interval between attempts.
+	// tink-worker is a daemon process, so we do not exit here.
+	var conn *grpc.ClientConn
+	for {
+		conn, err = client.EstablishTinkServerConnection(flagEnvSettings.TinkServerGRPCAuthority, creds)
+		if err == nil {
+			break
+		}
+
+		fmt.Println("Error establishing connection to tink-server:", err)
+		// sleep a randomized amount of time before reconnecting to avoid thundering herds
+		// TODO: we may want to make this configurable via a cmdline option?
+		rand.Seed(time.Now().UnixNano())
```

Comment:
One thing to keep in mind is if a bunch of instances of `tink-worker` start at the same time and fail to connect at the same time, then the same seed value will be used. There might be some other nuances about using the time as a seed that might influence the sleep in weird ways. In this format, the seed will be an ever growing number that might not be very different between executions of `client.EstablishTinkServerConnection`.

An alternative is to seed it once using a value generated with `crypto/rand`.

Reply:
Yes, in theory a bunch of tink-workers could run the Seed function the same time. But the `UnixNano()` function is getting resolution down to the nanosecond. I think it's a reasonable risk, and even the Go by Example website uses this, only recommending `crypto/rand` for secret values.

Reply:
In playing around with this pattern, I think the distribution of results isn't as much a concern I initially thought. However - `rand.Intn` can return 0. If the call to the EstablishTinkServerConnection function fails, there is a chance the loop will run again using the same Unix epoch nanosecond as the seed, which results in the same random value of 0 (for example, a seed of `1639594958949741000` always results in Intn returning 0).

Here's some janky test code I used to play around with this that doesn't involving actually sleeping:

```go
package main

import (
	"log"
	"math/rand"
	"time"
)

func main() {
	log.SetFlags(0)

	currentUnixNano := time.Now().UnixNano()
	var lastSleep time.Duration

	mean := 0
	sleepToNum := map[int]int{}
	execs := 100

	for i := 0; i < execs; i++ {
		// This comment represents a call to net.Dial / whatever.
		example()

		currentUnixNano += lastSleep.Nanoseconds()
		rand.Seed(currentUnixNano)
		sleepSeconds := rand.Intn(120)
		lastSleep = time.Duration(sleepSeconds)*time.Second
		mean += sleepSeconds
		v := sleepToNum[sleepSeconds]
		sleepToNum[sleepSeconds] = v+1
		log.Printf("seed: %d | sleepSeconds: %d | duration: %v",
			currentUnixNano, sleepSeconds, lastSleep)
	}

	mode := 0
	modeN := 0
	for sleep, occurrences := range sleepToNum {
		if occurrences > modeN {
			mode = sleep
			modeN = occurrences
		}
	}

	log.Printf("mean: %d | mode: %d | unique: %d | sleepToNum: %v",
		mean/execs, mode, len(sleepToNum), sleepToNum)
}

func example() {

}
```

Reply:
I appreciate your thorough explanations on these issues, they're very informative to me in this learning process. Since we're dealing with a network connection retry, I'm not too worried about a rare occurrence of an immediate retry (rand.Intn returning 0), and I doubt we'd make it through another reconnect attempt within the same nanosecond.